### ISSUE TITLE
Allocated Memory Exceede while training Fully Convolutional network for Image Segmentation

### ISSUE THREAD

**Issue by Gnopal1132** [tensorflow/models#9890]
I am trying to implement FCN from scratch on Pascal VOC 2012. But every time I try to train it memory goes out. I tried :

1. Reducing batch size to the minimum value of 4.
2. Reduce Data Augmentation.

I don't know what's causing the problem.

My Data Loader looks like this: 
1. Preprocess function takes the input tensor to apply data augmentation, resize the image to [224x224x3] and labels to [224x224x21]
![Unbenannt](https://user-images.githubusercontent.com/51056214/114296312-eb6f7480-9aaa-11eb-9737-1981acd0d59c.png)

My Model:
![Unbenannt](https://user-images.githubusercontent.com/51056214/114296389-79e3f600-9aab-11eb-9946-c957ef1d8e2b.png)

Executing the model:
![Unbenannt](https://user-images.githubusercontent.com/51056214/114296475-d0e9cb00-9aab-11eb-8b67-6441696c4319.png)

What should I do?

**Comment by Gnopal1132** [comment]
I gave my complete code above and I realized something: when training the model with a batch size of 20, I don't think so memory goes out, but while predicting on the validation set for that epoch the memory curve spikes very high and it goes out of the limit. What should I do?

**Comment by eirikaso** [comment]
Are you running out of GPU memory or RAM? I have issues that the RAM runs out when running evaluation.

**Comment by Gnopal1132** [comment]
The Ram actually runs out !!

**Comment by eirikaso** [comment]
OK, same issue here. I can run evaluation on a couple of checkpoints it seems, but the RAM usage seem to increase on every new iteration. For now I've just ran a single iteration and then killed the process a couple times during training but not exactly optimal. I think I'll have to make a script that restarts it after every checkpoint and see if that works or something

**Comment by eirikaso** [comment]
It seems now that the memory only builds up until a certain point for me and then stabilize. After adding a large SWAP partition of 64GB, I have no issues. My whole RAM gets filled up and a few GB of my SWAP. The training no longer fails, so I'm happy with that. 

Note that I also set the "num_replicas" argument to 0 in the "eager_train_step" function call earlier. I do not know if this also impacted since I have not reverted this change after adding more swap

**Comment by Petros626** [comment]
just reduce the batch_size and you're fine with training. Please close this issue, when you don't need further help

**Comment by PardeepGothriwal** [comment]
Is anyone currently working on this? If not, I’d like to take a stab at it and share my progress here.
